### ISSUE TITLE
twisted.mail.smtp.ESMTPClient tries to STARTTLS multiple times

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6524]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6524|
|Type|release blocker: regression|
|Created|2013-05-23 10:57:30Z|
|Branch|https://github.com/twisted/twisted/tree/judicious-smtp-starttls-6524|

If `ESMTPClient` is used to connect to an SMTP server that offers the `STARTTLS` feature, it will try to use it.  If the SMTP server still offers the `STARTTLS` after the command has been used once to start TLS, `ESMTPClient` will try to use it again.  This is useless at best, and in practice probably always triggers an error from the server and prevents mail from being sent.

This is a regression introduced by [#3989](https://github.com/twisted/twisted/issues/3989).

Attachments:

* [mailish.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94f/94f137ea09e6221da2c47b0874c99935494b8135/df25dc0c8898f4bf1193131b5dc8a1366d700ad2.py) (1178 bytes) - added by exarkun on 2013-05-23 11:01:00Z - It does not work.

<details><summary>Searchable metadata</summary>

```
trac-id__6524 6524
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_judicious_smtp_starttls_6524 branches/judicious-smtp-starttls-6524
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1369306650000000 1369306650000000
changetime__1371470604000000 1371470604000000
version__None None
owner__exarkun exarkun
cc__ashfall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Here's a transcript demonstrating the consequences of this bug:

```

<<< 220 email-smtp.amazonaws.com ESMTP SimpleEmailService-376766033
>>> EHLO top
<<< 250-email-smtp.amazonaws.com
<<< 250-8BITMIME
<<< 250-SIZE 10485760
<<< 250-STARTTLS
<<< 250-AUTH PLAIN LOGIN
<<< 250 Ok
>>> STARTTLS
<<< 220 Ready to start TLS
>>> EHLO top
<<< 250-email-smtp.amazonaws.com
<<< 250-8BITMIME
<<< 250-SIZE 10485760
<<< 250-STARTTLS
<<< 250-AUTH PLAIN LOGIN
<<< 250 Ok
>>> STARTTLS
<<< 454 TLS not available due to temporary reason: TLS already active
>>> QUIT
```

And attached is a script that reproduces it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [38818]) Branching to 'judicious-smtp-starttls-6524'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Relatively straightforward; only wished I had schematic tables once during the entire development process (arguably the minimum number of times it is possible to wish for schematic tables when writing a piece of software).

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @exarkun<br>@radix set status to `new`|
|-|-|

Schematic tables would be nice. Maybe we should sprint on them some time.

[1] You have a "verify that" in a docstring in a test module, but I guess that's ok because it's not a test_ method itself, just a utility assertion method?

No flakes, tests pass, +1, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [38833]) Merge judicious-smtp-starttls-6524

Author: exarkun
Reviewer: radix
Fixes: [#6524](https://github.com/twisted/twisted/issues/6524)

Avoid having `twisted.mail.smtp.ESMTPClient` try to use the STARTTLS command
if the server offers it but the connection is already using TLS.
